(WIP) Course Entrance Exams

mattdrayer opened the pull request:
Up for review is our new "Entrance Exams" feature!  There are a lot of parts comprising this functionality, so we'll try to keep track of the main points of reference in this first post.

*JIRA Epic*: https://openedx.atlassian.net/browse/SOL-40
*Solutions Review*: @chrisndodge 
*T&L Review*: @andy-armstrong 
*Sandbox*: http://studio.entexams.m.sandbox.edx.org/course/edX/DemoX/Demo_Course
*InvisionApp*: http://invis.io/T61UAL5UX
*A11y Review*:  @cptvitamin 
*HTML/CSS Review*: @marcotuts

chrisndodge commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -652,6 +663,31 @@ class CourseFields(object):
+        default=None,
+        scope=Scope.settings,
+    )
+
```
Will this Id field be visible in the Advanced Settings? If so, I think we'd want to hide this from uses directly editing it.

Also we should talk about export/import lifecycles (e.g. importing a course into a course with a different course_id) as I presume this pointer will need to get updated on import.

mattdrayer replied:
We're hiding the three new entrance_exam fields from the Advanced Settings screen via the FILTERED_LIST dictionary found in course_metadata.py

mattdrayer replied:
@marcotuts and I talked about import/export support for MVP and the decision was to tackle this functionality as part of the Q3 follow-on work

chrisndodge replied:
re: export/import, gotcha. In the meantime, I don't know if we should a) document this limitation or b) somehow explicitly error out (aka assert) so that we don't get authors into a bad state. I'd prefer a hard assertion rather than silent failures, but it wouldn't be a must have.

mattdrayer replied:
If you can show me where we'd insert the assertion I can add it in -- let me know

chrisndodge commented on lms/djangoapps/courseware/grades.py:
```diff
@@ -22,33 +22,12 @@
-    while len(stack) > 0:
-        next_descriptor = stack.pop()
-        stack.extend(get_dynamic_descriptor_children(next_descriptor))
-        yield next_descriptor
```
why the moving of this block of code?

mattdrayer replied:
I tried referencing yield_dynamic_descriptor_descendents directly but it resulted in a circular reference -- I ended up having to move the block to a separate location and pointing both modules to the new spot.

chrisndodge commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -776,7 +778,22 @@ def settings_handler(request, course_key_string):
+                        # setting to default value from cms configurations.
+                        entrance_exam_minimum_score_pct = float(settings.ENTRANCE_EXAM_MIN_SCORE_PCT)
+                        ee_min_score_pct = request.json.get('entrance_exam_minimum_score_pct', None)
+                        if ee_min_score_pct != '' and ee_min_score_pct is not None:
```
I think 'if ee_min_score_pct:' suffices here

chrisndodge commented on cms/djangoapps/contentstore/views/course.py:
```diff
@@ -776,7 +778,22 @@ def settings_handler(request, course_key_string):
+
+                        create_entrance_exam(request, course_key, entrance_exam_minimum_score_pct)
+                    else:
+                        delete_entrance_exam(request, course_key)
```
This goes back to some of my comments on the edx-milestones repo which is a product level decision regarding accidental removal of an extrance_exam and if all student achievements might be lost during a running course. I'll have to look at the front-end code, but we should make the author explicitly acknowledge that irreversable data loss will occur.

mattdrayer replied:
I'm 99% sure we'll be shifting edx-milestones over to a soft-delete approach, which should avoid the scenario you described -- let me know if additional work would be needed beyond that.

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+    First clean out any old entrance exams.
+    """
+    _delete_entrance_exam(request, course_key)
+    return _create_entrance_exam(request=request, course_key=course_key, entrance_exam_minimum_score_pct=entrance_exam_minimum_score_pct)
```
nitpick: this is a bit long line. Except for entrance_exam_minimum_score_pct, you don't need calling parameter names for the first two.

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+    return _create_entrance_exam(request=request, course_key=course_key, entrance_exam_minimum_score_pct=entrance_exam_minimum_score_pct)
+
+
+def _create_entrance_exam(request, course_key, entrance_exam_minimum_score_pct=50.0):
```
should the default value for _pct come from configuration rather than code?

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+    factory = RequestFactory()
+    internal_request = factory.post('/', json.dumps(payload), content_type="application/json")
+    internal_request.user = request.user
+    created_item = json.loads(create_item(internal_request).content)
```
Sigh. Wish create_item didn't need a request... Do you know why?

mattdrayer replied:
create_item() is a new accessor for the pre-existing internal _create_item() method, which pulls data from request.json in order to populate various fields for the item being created.  I didn't want to reference the private method directly in the entrance_exam module so opted for this approach.

cpennington replied:
I mentioned this over hipchat, but I would have preferred to extract the data-only portions of this (perhaps into an `api.py` module), and then call that directly, rather than manufacturing a fake request.

mattdrayer replied:
I like the idea of an api.py module, but where would it live?  Just alongside everything else in the contentstore/views folder?

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+        'entrance_exam_minimum_score_pct': entrance_exam_minimum_score_pct / 100,
+        'entrance_exam_id': created_item['locator'],
+    }
+    CourseMetadata.update_from_dict(metadata, course, request.user)
```
Hmm. these are xBlocks fields on course right? If so, I think it's a bit cleaner to just set the property accessors:

course.entrance_exam_enabled = True
: 
:

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+    CourseMetadata.update_from_dict(metadata, course, request.user)
+
+    # Add an entrance exam milestone if one does not already exist
+    namespace = '{}.entrance_exams'.format(unicode(course_key))
```
I presume the format of 'namespace' is shared between LMS and CMS. Is that right? If so, I think it'd be better to have a shared utility method in common so this knowledge doesn't have to be re-done on the LMS side.

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+        })
+    milestones_api.add_course_milestone(
+        unicode(course.id),
+        'requires',
```
As I was mentioning in the edx-milestones reviews, I think it would be good to have these 'system defined' types as constants in the edx-milestones app so we can just refer to the constant here.

mattdrayer replied:
I've changed all of the string references over to use a dict provided by the milestones app via the API -- take a look and let me know what you think.

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+        return HttpResponse(status=400)
+
+    children = modulestore().get_items(course_key, qualifiers={'metadata.is_entrance_exam': True})
+    if len(children):
```
nitpick: if children  will suffice here

mattdrayer replied:
Will address, thanks

chrisndodge commented on cms/djangoapps/contentstore/views/entrance_exam.py:
```diff
@@ -0,0 +1,206 @@
+            'entrance_exam_minimum_score_pct': None,
+            'entrance_exam_id': None,
+        }
+        CourseMetadata.update_from_dict(metadata, course, request.user)
```
likewise, I think we can use the normal xBlock field setters directly to make this a bit simplier

chrisndodge commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -479,6 +489,14 @@ def _create_item(request):
+            if category == 'chapter' and is_entrance_exam:
+                metadata['is_entrance_exam'] = is_entrance_exam
+                child_position = 0
+
```
Should we assert if is_entrance_exam is set and category != 'chapter'?

mattdrayer replied:
Hmm...there really isn't a way to introduce that scenario at the moment because it's all in the system.  If someone did end up modifying the system to support alternative module types being flagged as entrance exams then the tests would need to be modified accordingly to support the changes.  And if anything I think in that scenario we'd *remove* the chapter constraint.  So I guess it doesn't seem necessary to me -- thoughts?

chrisndodge commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -492,8 +510,32 @@ def _create_item(request):
+                        grading_model['type'],
+                        request.user
+                    )
+
```
seems like the setting of the grading policy should be outside of _create_item at a slightly higher level of the business logic. Likewise, if we delete an entrance exam, we'd want to make sure we remove this grader policy.

Also, does this new grading policy show up in Studio's grading setting page (sorry, I can't remember the exact name of the page), as we wouldn't want authors to edit/delete it.

mattdrayer replied:
I'll look into suppressing the Entrance Exam from the grading setting page.  It does show up in the list once set.

mattdrayer replied:
Okay, I was able to locate the spot in contentstore/views/item.py that controls the display of graders in the dropdown -- we now filter out 'Entrance Exam' graders from the set

chrisndodge commented on cms/djangoapps/models/settings/course_details.py:
```diff
@@ -39,6 +41,8 @@ def __init__(self, org, course_id, run):
         self.course_image_name = ""
         self.course_image_asset_path = ""  # URL of the course image
+        self.entrance_exam_enabled = ""  # is entrance exam enabled
+        self.entrance_exam_minimum_score_pct = "50"  # grade requirements for entrance exam
```
likewise, should we pull this default from configuration?

mattdrayer replied:
Yep, will update

chrisndodge commented on cms/djangoapps/models/settings/course_details.py:
```diff
@@ -161,7 +165,8 @@ def update_from_json(cls, course_key, jsondict, user):
         # to make faster, could compare against db or could have client send over a list of which fields changed.
         for attribute in ABOUT_ATTRIBUTES:
-            cls.update_about_item(course_key, attribute, jsondict[attribute], descriptor, user)
+            if attribute in jsondict:
```
why this new if statement?

mattdrayer replied:
@asadiqbal08 -- can you comment on this change?

asadiqbal08 replied:
@chrisndodge it is just a safe side check. 
IF we are assuming that jsondict always contains the attributes that are defined in ABOUT_ATTRIBUTES list then this extra check is not required otherwise there will be an exception.

chrisndodge replied:
Although - by eye - this looks ok, because this change is outside of a feature flag (and would affect all instances, including edx.org), a quick sign off from T&L might be a good idea: @cahrens @explorerleslie

mattdrayer replied:
@andy-armstrong -- just want to make sure this particular issue does not fall thru the cracks -- if you can take a look and let me know what you think, that would be great.

chrisndodge commented on cms/djangoapps/models/settings/course_metadata.py:
```diff
@@ -56,21 +62,28 @@ def fetch(cls, descriptor):
+        """
+        Fetches all key:value pairs from persistence and returns a CourseMetadata model.
+        """
+        result = {}
```
My guess here is that you needed a new entry point (fetch_all) to provide a non-filtered set, right?

mattdrayer replied:
Yes, that is correct -- the existing fetch() was filtering out the metadata because it's fairly coupled to the view case.

chrisndodge commented on common/lib/xmodule/xmodule/course_module.py:
```diff
@@ -202,6 +206,13 @@ class CourseFields(object):
+                    "min_count": 0,
+                    "drop_count": 0,
+                    "weight": 0.0,
+                },
```
I believe this will mean that all newly created courses will have this grading policy? I believe there is code which sets this when the author selects to have an entrance exam. Should we then leave it off of the default value?

mattdrayer replied:
I can remove it from here, yes -- I think this is a holdover from R&D work

chrisndodge commented on common/lib/xmodule/xmodule/modulestore/mongo/base.py:
```diff
@@ -1159,7 +1159,13 @@ def create_child(self, user_id, parent_usage_key, block_type, block_id=None, **k
+                parent.children.append(xblock.location)
+            else:
+                parent.children.insert(kwargs.get('position'), xblock.scope_ids.usage_id)
+
```
why would one add an xblock.location and the other xblock.scope_ids.usage_id. I'm guessing they are equivalent, but - if so - then I'd suggest unifying both cases here to make it more readible

mattdrayer replied:
Will change to 'location'

chrisndodge commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -377,6 +458,13 @@ def handle_grade_event(block, event_type, event):
         dog_stats_api.increment("lms.courseware.question_answered", tags=tags)
 
+        # Fulfill any milestones for this content module now that all the data has been saved
+        _fulfill_content_milestones(
```
For performance reasons, ideally we'd only call this if the graded item was within an entrance exam. Thus my suggestion to make is_entrance_exam be 'inherited metadata' so that the setting on the Chapter would auto-magically appear on all of the content subtree.

chrisndodge replied:
Also, I know we are checking feature flags inside of _fulfill_content_milestones. If we add another FEATURE flag check here, it might help code readers know that it really only does anything with that feature flag on.

mattdrayer replied:
I'll add in the checks to minimize the scenarios where we do the noted work.

mattdrayer replied:
One comment here -- I think it's important to discern between fulfilling of milestones and the entrance exam case which leads to the fulfilling of milestones.  I think the outer check should be to see if the milestones app is available, and if so we step into the method and begin working through the various fulfillment scenarios.  I'll write it up in the code so you can see what I mean.

chrisndodge commented on common/lib/xmodule/xmodule/seq_module.py:
```diff
@@ -45,6 +45,16 @@ class SequenceFields(object):
+        ),
+        scope=Scope.settings,
+    )
+
```
We had briefly discussed making this being "inheritable metadata" to optimize the performance around determine if a setting of a score on a problem is within an entrance_exam, and - thus - need to compute if a milestone fulfillment has been achieved.

Maybe "in_entrance_exam" might be a slightly better name if we decide to use metadata inheritance.

mattdrayer replied:
Happy to employ inherited metadata -- let's discuss how to go about doing so (an example of it in action would be fantastic)

chrisndodge commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -345,7 +381,52 @@ def inner_get_module(descriptor):
+                    content_milestones = milestones_api.get_course_content_milestones(
+                        course_key,
+                        content_key,
+                        relationship='fulfills'
```
likewise, re: use of hard coded string constants

chrisndodge replied:
Also, I'm pretty sure the Analytics team is going to want to have an analytics event fired here. You should check in with them. @brianhw can you - or someone else - weigh in on this?

chrisndodge commented on lms/djangoapps/courseware/module_render.py:
```diff
@@ -345,7 +381,52 @@ def inner_get_module(descriptor):
+                    user = {'id': user_id}
+                    for milestone in content_milestones:
+                        milestones_api.add_user_milestone(user, milestone)
+
```
I don't know if we want to add another check if exam_pct < course.entrance_exam_minimum_score_pct *AND* the student has received the milestone already, then we remove it. I'm thinking of cases where perhaps an author adds more problems to an exam during a course run. 

It comes down to a question of whether achieved milestones can be revoked if the criteria changes. @marcotuts?

marcotuts replied:
@chrisndodge In discussing this with @Matt previously, we landed in a place where changes to a course exam wouldn't necessary revoke a milestone. This is one of the benefits of separating out milestones from the success criteria elements/requirements. Milestone requirements may change in the future but students shouldn't necessarily lose access to content based on exam content changes.

chrisndodge commented on lms/djangoapps/courseware/tabs.py:
```diff
@@ -0,0 +1,55 @@
+    # If the course has an entrance exam, we'll need to see if the user has not passed it
+    # If so, we'll need to hide away all of the tabs except for Courseware and Instructor
+    entrance_exam_mode = False
+    if settings.FEATURES.get('ENTRANCE_EXAMS', False) and settings.FEATURES.get('MILESTONES_APP', False):
```
This is probably a nitpick, but while I'm looking at the code, I'd suggest collapsing these. ENTRANCE_EXAMS must depend on MILESTONES_APP, so I think it's redundant to have two feature flags.

mattdrayer replied:
Yep, makes sense, i'll take care of that

chrisndodge commented on lms/djangoapps/courseware/tabs.py:
```diff
@@ -0,0 +1,55 @@
+    # If so, we'll need to hide away all of the tabs except for Courseware and Instructor
+    entrance_exam_mode = False
+    if settings.FEATURES.get('ENTRANCE_EXAMS', False) and settings.FEATURES.get('MILESTONES_APP', False):
+        if getattr(course, 'entrance_exam_enabled', False):
```
I've seen this in a few places, but since 'entrance_exam_enabled' is an xBlock field, it'll always be an attribute so you don't have to do a getattr()

mattdrayer replied:
What about for pre-existing content?  I've done it this way because it didnt' seem to be present for courses which hadn't had the flag set via the new workflow.

chrisndodge commented on lms/djangoapps/courseware/tabs.py:
```diff
@@ -0,0 +1,55 @@
+        if getattr(course, 'entrance_exam_enabled', False):
+            course_milestones_paths = get_course_milestones_fulfillment_paths(
+                unicode(course.id),
+                user.__dict__
```
hmm, this seems like odd syntax. Should we have an explicit utility method in milestones_app to covert to a dict?

mattdrayer replied:
I'll add a function to milestones_helpers.py to handle this

chrisndodge commented on lms/djangoapps/courseware/tabs.py:
```diff
@@ -0,0 +1,55 @@
+            if tab.type not in ['courseware', 'instructor']:
+                continue
+            if tab.type == 'courseware':
+                tab.name = "Entrance Exam"
```
Sorry, I can't remember if tab.name is a human visible string. If so, please I18N.

chrisndodge replied:
Is tab.name a human visible string?

andy-armstrong replied:
Yes, I think this needs i18n.

mattdrayer replied:
Can I see an example of how we apply i18n at the app/python tier?

marcotuts replied:
This may just be a matter of wrapping the string with ugettext("String") and adding the import at the top of the file: 

"from django.utils.translation import ugettext"

(references:
I went from here: https://github.com/edx/edx-platform/wiki/i18n-Coding-Guidelines
to here: https://github.com/edx/edx-platform/search?l=python&q=ugettext%28&type=Code&utf8=%E2%9C%93)

andy-armstrong replied:
I still think i18n needs to be applied here.

mattdrayer replied:
I've added this just now -- missed it in all the commotion :)

On Wed, Jan 14, 2015 at 2:28 PM, Andy Armstrong <notifications@github.com>
wrote:

> In lms/djangoapps/courseware/tabs.py
> <https://github.com/edx/edx-platform/pull/6455#discussion-diff-22962650>:
>
> > +                    for content in value['content']:
> > +                        if content['content_id'] == course.entrance_exam_id:
> > +                            entrance_exam_mode = True
> > +                            break
> > +
> > +    # Now that we've loaded the tabs for this course, perform the Entrance Exam mode work
> > +    # Majority case is no entrance exam defined
> > +    course_tab_list = []
> > +    for tab in xmodule_tab_list:
> > +        if entrance_exam_mode:
> > +            # Hide all of the tabs except for 'Courseware' and 'Instructor'
> > +            # Rename 'Courseware' tab to 'Entrance Exam'
> > +            if tab.type not in ['courseware', 'instructor']:
> > +                continue
> > +            if tab.type == 'courseware':
> > +                tab.name = "Entrance Exam"
>
> I still think i18n needs to be applied here.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6455/files#r22962650>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

chrisndodge commented:
@mattdrayer @smagoun @marcotuts done with review. Main concerns to investigate:

1) Will the "Entrance Exam" grading policy be visible/editable/deleteable in Studio Grading Settings page? If so, we should hide it from that page so that authors do not manipulate those system defined policy settings

2) If we can 'quick exit' in the on grading event so that we don't recompute the grade on the entrance exam when someone answers a problem that is outside of the exam. There's a suggestion to leverage the existing concept of 'inherited metadata'.

Other than that, a few stylistic issues.

marcotuts commented:
@cptvitamin - Hi Mark, 

This is a WIP PR for the Entrance Exams feature, which effectively gates access to the rest of a course until a student has completed the entrance exam. Would love to get feedback on A11y and find any improvements we can make to this feature. You can try it out on the active sandbox, and if you have any questions I'm happy to walk through it in person as well.

(log in as staff) Configuration: http://studio.entexams.m.sandbox.edx.org/settings/details/edX/DemoX/Demo_Course
(log in as staff) Authoring: http://studio.entexams.m.sandbox.edx.org/course/edX/DemoX/Demo_Course
(log in as audit) Student Experience:   http://entexams.m.sandbox.edx.org/courses/edX/DemoX/Demo_Course/courseware/8892878164ee4267b80d9ab4743373d7/49e21d21b72f4772918908edd9200e1f/

mattdrayer commented:
@chrisndodge -- most of the feedback for this PR has been addressed at this point -- keeping my eye on Jenkins to see how the tests go -- if you have some time tomorrow it would be great if you can point me to an example of 'inherited metadata', which we can use to 'quick exit' on the grading event when the module is not part of an entrance exam.

mattdrayer commented:
@chrisndodge -- I've implemented 'inherited metadata' as we discussed and added the check to the grading logic for quick-exiting.  In addition I was able to get to the bottom of the failing bok choy tests (@asadiqbal08, FYI they are working now).  I'm going to provision a new entexams sandbox for @cahrens to look at in the morning -- let me know if you see anything else that needs to be addressed.

chrisndodge commented on lms/djangoapps/courseware/tests/test_entrance_exam.py:
```diff
@@ -0,0 +1,276 @@
+            display_name="Exam Sequential - Subsection 1",
+            graded=True,
+            metadata={'in_entrance_exam': True}
+        )
```
I think if we want to test that the "quick exit" regarding when to recalculate score averages is working with the new inherited metadata, we'd also like a negative case: a sequential which does not have 'in_entrance_exam' set to false and has a problem. Then the test assertion would to make sure that submitting an answer on the problem that is not in an entrance exam does not call the score calculation function.

chrisndodge commented:
Just a quick question on whether tabs.name should be I18N'd.

Let's move this review onto a 2nd reviewer from T&L. FYI: @cahrens @explorerleslie

cahrens commented:
Ticket for TNL review is https://openedx.atlassian.net/browse/TNL-1161.

Please add Jasmine tests for JS changes, and a smoke-level bok choy test. Ideally those should be done before we review, but given the deadline, we will start reviewing while the tests are being written.

cahrens commented:
@andy-armstrong Here is the PR.
Sandbox is  http://prereqs.m.sandbox.edx.org

chrisndodge commented:
FYI, PEP8 violation is (which diff_quality can't pick up):

13:23:17 pep8 common | tee /home/jenkins/workspace/edx-all-tests-auto-pr/SHARD/1/TEST_SUITE/quality/reports/common/pep8.report
13:23:17 common/lib/xmodule/xmodule/modulestore/inheritance.py:184:1: E302 expected 2 blank lines, found 1
13:23:17 Number of pep8 violations: 1

andy-armstrong commented:
FYI I'm going to be the T&L reviewer on this PR.

andy-armstrong commented:
There is quite a bit of uncovered JavaScript according to diff-cover:

https://jenkins.testeng.edx.org/job/edx-all-tests-auto-pr/21586/SHARD=1,TEST_SUITE=unit/artifact/reports/diff_coverage_combined.html

Please bear this in mind when you create the Jasmine tests.

andy-armstrong commented:
Can you point me at the JIRA story for this feature, along with any documentation that explain the functionality. Please update the PR and the commit message to include the JIRA ticket number too. Thanks.

chrisndodge commented:
Note @andy-armstrong that @marcotuts is the Product Owner if you have any questions.

mattdrayer commented:
@andy-armstrong -- I'll be addressing the JS test coverage today

andy-armstrong commented:
What does the clock icon mean here? Is this a standard LMS convention? It could use a tooltip.

![image](https://cloud.githubusercontent.com/assets/5985072/5723339/64337c72-9b11-11e4-8b31-f51cbce08e9d.png)

andy-armstrong commented:
As "Staff", when I click on "Instructor" then all the tabs reappear and I can click on them. So then I can get to the discussion, the course info etc:

http://entexams.m.sandbox.edx.org/courses/AndyA/ATC/1/courseware/2e12a2ebb5ad4ce6b856933feee8c7b3/

I don't imagine this is a showstopper ("Well, don't click on it then!") but I figured I'd report it.

andy-armstrong commented:
The student can still navigate to course pages before they've passed the exam if they have the URL. For example, I was able to get to the discussion when logged in as 'honor@example.com':

http://entexams.m.sandbox.edx.org/courses/AndyA/ATC/1/discussion/forum

![image](https://cloud.githubusercontent.com/assets/5985072/5723497/669cf500-9b12-11e4-8205-875169a9d8e0.png)

marcotuts commented:
@andy-armstrong, @mattdrayer 
A couple notes Andy in response to your initial comments/screenshots:
1.) The icon on the LMS is an existing pattern used to designate a "graded" subsection. The lack of tooltip is an ancient but lingering bug in a backlog somewhere. I'd looove to fix it but we can address in a separate PR. :) 
2.) For staff members we may actually want to keep a version of this behavior and even extend the recent work done on the instructor preview bar to let an instructor 'preview' their course as a student who has completed or not completed the entrance exam milestone. I can create a follow up story for early Q3 to review this work. Thanks for noting it here!  Story - https://openedx.atlassian.net/browse/SOL-215
3.) This is a known issue that @mattdrayer and @chrisndodge discussed adding to the PR. Blocking access to direct url access would be a helpful addition. If for whatever reason we wind up punting on this addition to this PR, let me know and I can create another story, but ideally we could include this in the current PR. 

Thanks!

andy-armstrong commented:
I'm not sure if I've got my course in a strange state, but "View Course" in the LMS it taking me to the course info page now:

1. Sign in as honor@example.com here: http://entexams.m.sandbox.edx.org/
2. Click on "View Course" for AndyA's Test Course

You'll end up on this info page:

http://entexams.m.sandbox.edx.org/courses/AndyA/ATC/1/info

There doesn't seem to be a way to get to the entrance exam. If I click on the "Entrance Exam" button I end up on a very blank page:

![image](https://cloud.githubusercontent.com/assets/5985072/5726363/a18528f2-9b28-11e4-9a7f-d1cd7ce48f60.png)

mattdrayer commented:
@andy-armstrong  -- the 'very blank page' you are seeing is due to having no Unit, etc. defined and published for the course in question

andy-armstrong commented:
@mattdrayer Ah, of course! I've published it now and now it shows the first unit in the course. It seems that I've skipped over the entrance exam. Why can't I get back to the entrance exam since I haven't passed it yet? Or if I have passed it, how come I don't have all the other tabs.

andy-armstrong commented:
@mattdrayer Sorry, I see that I didn't put any content in the entrance exam either. Now that I have some it works as I would have expected.

srpearce commented on cms/templates/js/mock/mock-settings-page.underscore:
```diff
@@ -62,6 +62,21 @@
+            <div class="show-data">
+                <div class="heading">
+                    <input type="checkbox" id="entrance-exam-enabled" />
+                    <label for="entrance-exam-enabled">${_("Enables entrance exam requirement. Students must successfully complete exam before gaining access to the course.")}</label>
```
Text for this check box:

Require students to pass an exam before beginning the course.

mattdrayer replied:
Updated (code to be pushed momentarily)

srpearce commented on cms/templates/js/mock/mock-settings-page.underscore:
```diff
@@ -62,6 +62,21 @@
+                    <label for="entrance-exam-enabled">${_("Enables entrance exam requirement. Students must successfully complete exam before gaining access to the course.")}</label>
+                </div>
+                <div id="div-grade-requirements">
+                    <p><span class="tip tip-inline">${_("You can now view and author your course entrance exam from the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a></span></p>
```
To create your course entrance exam, go to the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a>. An Entrance Exam section will be created automatically.

mattdrayer replied:
Updated (code to be pushed momentarily)

srpearce commented on cms/templates/js/mock/mock-settings-page.underscore:
```diff
@@ -62,6 +62,21 @@
+                </div>
+                <div id="div-grade-requirements">
+                    <p><span class="tip tip-inline">${_("You can now view and author your course entrance exam from the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a></span></p>
+                    <p><label>${_("Grade Requirements")}</label></p>
```
Minimum Passing Score

cptvitamin replied:
this ```<label>``` has no ```for``` attribute, which is *required* unless the form input it references is nested inside of it.  Should be ```<label for="entrance-exam-minimum-score-pct">```

mattdrayer replied:
Updated (code to be pushed momentarily)

srpearce commented on cms/templates/js/mock/mock-settings-page.underscore:
```diff
@@ -62,6 +62,21 @@
+                <div id="div-grade-requirements">
+                    <p><span class="tip tip-inline">${_("You can now view and author your course entrance exam from the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a></span></p>
+                    <p><label>${_("Grade Requirements")}</label></p>
+                    <p><div><input type="text" id="entrance-exam-minimum-score-pct"><span class="tip tip-inline">${_("as a percent, e.g. 50")}</span></div></p>
```
I'd just have the text box with a percent sign next to it here (don't need the explanatory text).

cptvitamin replied:
whether you go with @srpearce's suggestion or not, that text or the text that is there needs to be programmatically associated with the form input with aria-describedby: 
```<input type="text" id="entrance-exam-minimum-score-pct" aria-describedby="min-score-format"><span id="min-score-format" class="tip tip-inline">${_("as a percent, e.g. 50")}</span>```
also, having a ```<div>``` nested inside a ```<p>``` is invalid HTML

mattdrayer replied:
Updated (code to be pushed momentarily)

srpearce commented on cms/templates/js/mock/mock-settings-page.underscore:
```diff
@@ -62,6 +62,21 @@
+                    <p><span class="tip tip-inline">${_("You can now view and author your course entrance exam from the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a></span></p>
+                    <p><label>${_("Grade Requirements")}</label></p>
+                    <p><div><input type="text" id="entrance-exam-minimum-score-pct"><span class="tip tip-inline">${_("as a percent, e.g. 50")}</span></div></p>
+                    <p><span class="tip tip-inline">${_("The score student must meet in order to successfully complete the entrance exam. ")}</span></p>
```
The minimum score a student must receive to pass the entrance exam.

cptvitamin replied:
This text should also be programmatically associated with that form field.  luckily aria-describedby takes a space-separated list of IDrefs:
```<input type="text" id="entrance-exam-minimum-score-pct" aria-describedby="min-score-format min-score-tip"><span id="min-score-format" class="tip tip-inline">${_("as a percent, e.g. 50")}</span>```
...
```<span class="tip tip-inline" id="min-score-tip">${_("The score student must meet in order to successfully complete the entrance exam. ")}</span>```

mattdrayer replied:
Updated (code to be pushed momentarily)

srpearce commented on cms/templates/settings.html:
```diff
@@ -308,6 +308,24 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                    <div class="show-data">
+                        <div class="heading">
+                            <input type="checkbox" id="entrance-exam-enabled" />
+                            <label for="entrance-exam-enabled">${_("Enables entrance exam requirement. Students must successfully complete exam before gaining access to the course.")}</label>
```
See suggestions for mock-settings-page.underscore file above

mattdrayer replied:
Updated (code to be pushed momentarily)

andy-armstrong commented on cms/djangoapps/contentstore/views/tests/test_entrance_exam.py:
```diff
@@ -0,0 +1,248 @@
+    from util.milestones_helpers import serialize_user
+
+
+class EntranceExamHandlerTests(CourseTestCase):
```
Awesome tests!

mattdrayer replied:
There are more coming! :)

andy-armstrong commented on cms/static/js/models/settings/course_details.js:
```diff
@@ -43,6 +45,12 @@ var CourseDetails = Backbone.Model.extend({
         }
+        if(_.has(newattrs, 'entrance_exam_minimum_score_pct')){
+            var intEntranceExamMinScore = Math.round(newattrs.entrance_exam_minimum_score_pct); // see if this ensures value saved is int
+            if (!isFinite(intEntranceExamMinScore) || /\D+/.test(newattrs.entrance_exam_minimum_score_pct) || intEntranceExamMinScore < 0 || intEntranceExamMinScore > 100) {
```
I think this would be easier to read separated out into a helper method. As it stands I found it hard to be sure that the logic was correct. In fact, I could see this being a beneficial method in a number of places where we validate string values that should be integers.

FYI, I tested the field manually and all the edge cases I expected worked. I was surprised that 0% was a valid value but I figure that there are use cases where that might be useful.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/models/xblock_info.js:
```diff
@@ -120,7 +120,11 @@ function(Backbone, _, str, ModuleUtils) {
+            /**
+             * True if this xblock contains an entrance exam module.
+             */
+            'is_entrance_exam': null
```
This seems like a strangely specific field to add to xblock info. If every feature adds a new field this is going to get out of hand (xblock info is already bloated). Was it not possible to make this a new xblock type? Could we have a new 'override_type' or something that returns an enum where 'entrance_exam' is one of the options?

I wouldn't hold up merging for this, but just wanted to record my thoughts.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/spec/views/settings/main_spec.js:
```diff
@@ -56,8 +58,34 @@ define([
             );
         });
+
+        it('Creating an entrance exam with minimum_score_pct. It should save the ' +
```
This name is too long! I'd go with something shorter that expresses the intent without all the details. Something like:

    it('should save entrance exam information correctly', function () {

andy-armstrong replied:
Note that the test name is meant to read as a sentence starting with the 'it'. I see that the pre-existing test in this file doesn't follow this convention either.

mattdrayer replied:
Updated (code to be pushed momentarily)

andy-armstrong commented on cms/static/js/spec/views/settings/main_spec.js:
```diff
@@ -56,8 +58,34 @@ define([
+            AjaxHelpers.expectJsonRequest(
+                requests, 'POST', urlRoot, expectedJson
+            );
+            AjaxHelpers.respondWithJson(requests, expectedJson);
```
I find it a little odd to send a response but then not verify that the UI responds to it correctly. Generally I'd have one test for a good response, and one for a server error. However, really that should be done in a general 'save settings' test, so I won't ask you to do that.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/spec/views/settings/main_spec.js:
```diff
@@ -56,8 +58,34 @@ define([
+                requests, 'POST', urlRoot, expectedJson
+            );
+            AjaxHelpers.respondWithJson(requests, expectedJson);
+        });
```
Please add a test for invalid inputs that verifies that a message is shown and that no AJAX request is made.

andy-armstrong replied:
Also add tests for checking and unchecking the 'enable exam' check box which verify that the correct request is made to the server.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/views/settings/main.js:
```diff
@@ -63,6 +63,15 @@ var DetailsView = ValidatingView.extend({
         var imageURL = this.model.get('course_image_asset_path');
         this.$el.find('#course-image-url').val(imageURL);
         this.$el.find('#course-image').attr('src', imageURL);
+        if(this.model.get('entrance_exam_enabled') == 'true'){
```
Why is this modeled as a string and not a boolean?

andy-armstrong replied:
Would you mind adding spaces around the parentheses like we do everywhere else. I don't know if this will fail our JSHint file or not (probably not), but it would be preferable to be be consistent within this code.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/views/settings/main.js:
```diff
@@ -63,6 +63,15 @@ var DetailsView = ValidatingView.extend({
         this.$el.find('#course-image').attr('src', imageURL);
+        if(this.model.get('entrance_exam_enabled') == 'true'){
+            this.$el.find('#' + this.fieldToSelectorMap['entrance_exam_enabled']).attr('checked', this.model.get('entrance_exam_enabled'));
+            this.$el.find('#div-grade-requirements').show();
```
We prefer ``this.$(`` as the standard shorthand for ``this.$el.find(``.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/views/settings/main.js:
```diff
@@ -151,6 +162,22 @@ var DetailsView = ValidatingView.extend({
             break;
+        case 'entrance-exam-enabled':
+            if($(event.currentTarget).is(":checked")){
+                this.$el.find('#div-grade-requirements').show();
```
We have a policy of not using ids unless they are absolutely necessary. I'd make this be a class instead.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/static/js/views/settings/main.js:
```diff
@@ -151,6 +162,22 @@ var DetailsView = ValidatingView.extend({
+        case 'entrance-exam-minimum-score-pct':
+            // If the val is an empty string then update model with default value.
+            if($(event.currentTarget).val() == ''){
+                this.model.set('entrance_exam_minimum_score_pct', this.model.defaults.entrance_exam_minimum_score_pct);
```
It would be good to add a unit test to verify that this defaulting works as expected.

mattdrayer replied:
@asadiqbal08, FYI -- this may be covered by the recent set of JS tests that we merged into the branch, however there are still 5-6 lines outstanding for main.js, according to the diff-coverage-combined report:  https://build.testeng.edx.org/job/edx-platform-report/29/Diff_Coverage_Report/

andy-armstrong commented on cms/static/js/views/settings/main.js:
```diff
@@ -151,6 +162,22 @@ var DetailsView = ValidatingView.extend({
+            break;
+        case 'entrance-exam-minimum-score-pct':
+            // If the val is an empty string then update model with default value.
+            if($(event.currentTarget).val() == ''){
```
We always use ``===`` (this will fail our JSHint validation).

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented on cms/templates/js/course-outline.underscore:
```diff
@@ -78,12 +78,14 @@ if (xblockInfo.get('graded')) {
-                        <span class="sr action-button-text"><%= gettext('Delete') %></span>
-                    </a>
-                </li>
+                <% if (!(xblockInfo.isEntranceExam())) { %>
```
I'd rather introduce a new method such as ```isReadOnly``` or ```canBeDeleted```. It doesn't seem right that entrance exams have to be handled specially in the template.

mattdrayer replied:
@asadiqbal08, FYI

andy-armstrong commented:
I've finished my first review. This is looking excellent, and with a few more tests I think it is ready to go. I have some stylistic comments that I'd prefer be addressed as well, but given the time crunch I won't be a stickler for those. Nice work guys!

mattdrayer commented:
Thanks Andy!  We'll address the items you've highlighted and will let you
know when the next revision/sandbox is available -- @asadiqbal08, just a
heads-up on the javascript items Andy has mentioned.

On Tue, Jan 13, 2015 at 4:14 PM, Andy Armstrong <notifications@github.com>
wrote:

> I've finished my first review. This is looking excellent, and with a few
> more tests I think it is ready to go. I have some stylistic comments that
> I'd prefer be addressed as well, but given the time crunch I won't be a
> stickler for those. Nice work guys!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/edx/edx-platform/pull/6455#issuecomment-69821314>.
>



-- 
Matthew Drayer
Solutions Architect
edX | MIT
mattdrayer@edx.org

cptvitamin commented on cms/templates/js/course-outline.underscore:
```diff
@@ -78,12 +78,14 @@ if (xblockInfo.get('graded')) {
+                <% if (!(xblockInfo.isEntranceExam())) { %>
+                    <li class="action-item action-delete">
+                        <a href="#" data-tooltip="<%= gettext('Delete') %>" class="delete-button action-button">
+                            <i class="icon fa fa-trash-o"></i>
```
I like to add an ```aria-hidden="true"``` to icon fonts as they can be focused on but don't speak anything by some screen readers.  They also render as a generic glyph in "visual" screen readers for people with cognitive disabilities. 
```<i class="icon fa fa-trash-o" aria-hidden="true"></i>```

mattdrayer replied:
Updated (code to be pushed momentarily)

cptvitamin commented on cms/templates/js/mock/mock-settings-page.underscore:
```diff
@@ -62,6 +62,21 @@
       </div>
     </li>
+    <li>
+        <label id="heading-entrance-exam">${_("Entrance Exam")}</label>
```
the ```<label> ``` element requires a ```for``` attribute which makes it inappropriate for grouping and labeling the inputs that follow.  I think the text "Entrance Exam" would be more useful as an ```<h3>``` here.

mattdrayer replied:
Updated (code to be pushed momentarily)

cptvitamin commented on cms/templates/settings.html:
```diff
@@ -308,6 +308,24 @@ <h2 class="title-2">${_("Requirements")}</h2>
                 </li>
+                % if settings.FEATURES.get('ENTRANCE_EXAMS'):
+                <li>
+                    <label id="heading-entrance-exam">${_("Entrance Exam")}</label>
```
[See previous comment](https://github.com/edx/edx-platform/pull/6455/files?diff=unified#r22899886)

mattdrayer replied:
Updated (code to be pushed momentarily)

cptvitamin commented on cms/templates/settings.html:
```diff
@@ -308,6 +308,24 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                        </div>
+                         <div id="div-grade-requirements">
+                          <p><span class="tip tip-inline">${_("You can now view and author your course entrance exam from the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a></span></p>
+                          <p><label>${_("Grade Requirements")}</label></p>
```
[See previous comments](https://github.com/edx/edx-platform/pull/6455/files?diff=unified#r22900656)

mattdrayer replied:
Updated (code to be pushed momentarily)

cptvitamin commented on cms/templates/settings.html:
```diff
@@ -308,6 +308,24 @@ <h2 class="title-2">${_("Requirements")}</h2>
+                         <div id="div-grade-requirements">
+                          <p><span class="tip tip-inline">${_("You can now view and author your course entrance exam from the ")}<a href='${course_handler_url}'>${_("Course Outline")}</a></span></p>
+                          <p><label>${_("Grade Requirements")}</label></p>
+                          <p><div><input type="text" id="entrance-exam-minimum-score-pct"><span class="tip tip-inline">${_("as a percent, e.g. 50")}</span></div></p>
```
[see comment](https://github.com/edx/edx-platform/pull/6455/files?diff=unified#r22900911)

mattdrayer replied:
Updated (code to be pushed momentarily)

marcotuts commented:
Hi everyone, this is a quick update of the conversations @srpearce and I had today around the learner's experience with entrance exams. We've created separate follow-on stories for these changes, but this message is here to make clear to reviewers what is included in this version of the feature. We expect that some of the current limitations will be resolved through these stories in upcoming sprints.

1.) A course about page doesn't currently include a note for courses that require an entrance exam, which would be a useful addition. A separate story has been created for this and added to the backlog here: https://openedx.atlassian.net/browse/SOL-218 

2.) Once a user enters a course with an active entrance exam they land on the Course Info page as they always do in the LMS. They must then realize that the entrance exam is available and click it in the header to complete the requirement gating the rest of the course. Removing this awkward half-step is described in this separate story: https://openedx.atlassian.net/browse/SOL-219 Other backlog stories will help this work as well, specifically - https://openedx.atlassian.net/browse/SOL-139) 

3.) Add message area above the entrance exam sequential describing your current state and score. Without this message it is unclear for students whether or not they have met the requirement without a page refresh. Here is the separate backlog story for this: https://openedx.atlassian.net/browse/SOL-221

4.) Explore options for asynchronously updating the score/state of the entrance exam to make clear to users what their progress is, as well as whether or not they have met the required score. Story: https://openedx.atlassian.net/browse/SOL-222

5.) Entrance Exam remediation for students who don't meet the required grade threshold is an experience improvement split off into this feature: https://openedx.atlassian.net/browse/SOL-220

Thanks!

andy-armstrong commented:
Sorry that I didn't notice this yesterday, but...

There is no way for an author to view/preview their course on LMS right now without passing the entrance exam themselves! I think there needs to be access control so that a staff user can preview/view live everything. The access control should only be for students, or for authors masquerading as one.

andy-armstrong commented:
"View Course" is still going to the course info page. Maybe it would be better if this page was accessible to students before they pass the exam. Is there material on there that they shouldn't be seeing?

andy-armstrong commented:
Never mind. The Course Info page shows the handouts, course news etc. I think there has to be some logic so that attempts to access the course info page will get redirected to the courseware page showing the entrance exam.

andy-armstrong commented on cms/djangoapps/contentstore/views/item.py:
```diff
@@ -786,6 +852,11 @@ def safe_get_username(user_id):
 
+    #instead of adding a new feature directly into xblock-info, we should add them into override_type.
+    override_type = {}
+    if getattr(xblock, "is_entrance_exam", None):
```
This isn't really what I was thinking. An 'override_type' would be a single string representing the true type of the object (i.e. 'entrance_exam') rather than what the xblock type is (i.e. 'chapter'). This seems like a more generic way to provide additional fields to xblock info which I think is too flexible.

I can let this go for the sake of getting the release out, but I'd like to see this done more cleanly.

andy-armstrong commented on cms/static/js/spec/views/settings/main_spec.js:
```diff
@@ -56,8 +58,48 @@ define([
+                .trigger('change');
+            expect(this.view.$('.div-grade-requirements div')).toBeVisible();
+
+            //input some invalid values.
```
I prefer to have lots of short tests, rather than one large test like this. I'd break out tests such as

    it('should disallow save with an invalid minimum score percentage'
    it('should provide a default value for the minimum score percentage'
    it('hides the entrance exam when the check box is deselected'

Usually I find that if my test needs comments breaking it into sections, then it is better to make each section its own test. It is then much clearer what has failed, and it is easier to run just the failing code.

andy-armstrong commented on common/test/acceptance/tests/lms/test_lms.py:
```diff
@@ -680,3 +681,73 @@ def test_python_execution_in_problem(self):
+        # Auto-auth register for the course
+        AutoAuthPage(self.browser, course_id=self.course_id).visit()
+
+    def test_entrance_exam_section(self):
```
I think you should have several Bok Choy tests for the different scenarios:

1. As a student, I can't enter the course before completing the entrance exam
2. As a student, the course becomes available to me as soon as I complete the entrance exam
3. As a student, I am shown an error message if my entrance exam score is too low

I'm sure there are others, but those are the ones that occur to me right now.

andy-armstrong commented:
I've finished my second review. I have a number of small concerns but nothing that should hold it up from merging. The biggest issue I see is the 'View Changes" button taking the user to the course info page. If that is considered okay, then that's fine with me as long as there is a story to fix it in the future.

:+1: once there is a clean build.

smagoun commented:
@marcotuts can you respond to Andy's comment about the "View Changes" button?

mattdrayer commented:
FYI, @chrisndodge and I discussed and have an angle for the "View Changes' button, but I'm not sure how long it will take to implement.

marcotuts commented:
@andy-armstrong and I resolved all open questions from a product standpoint and ensured open issues were captured as specific acceptance criteria on the follow-on Q3 stories.

chrisndodge commented:
I can sketch something out for you in a few hours as a pr onto your branch. Can you point me to where you are doing the check currently in the lms?

Sent from my iPhone

On Jan 14, 2015, at 4:59 PM, Matt Drayer <notifications@github.com> wrote:

> FYI, @chrisndodge and I discussed and have an angle for the "View Changes' button, but I'm not sure how long it will take to implement.
> 
> —
> Reply to this email directly or view it on GitHub.
>

andy-armstrong commented:
FYI, I mistakenly called it "View Changes" once (I've been doing preview too much the past few weeks). The issue is with the "View Course" button. Sorry to add unnecessary extra confusion.

@marcotuts and I both think that it is okay to merge the feature with the broken behavior and then fix it right afterwards.

jzoldak commented:
Also OK to merge with the new bok-choy test for test_page_has_enable_entrance_exam_field marked to skip and fix in the next sprint.

chrisndodge commented:
@mattdrayer bring on down to mergetown!

mattdrayer commented:
Ok, the quality, unit, acceptance and (pertinent) bok choy tests have finally passed the build -- thanks much to ALL -- we're merging!
